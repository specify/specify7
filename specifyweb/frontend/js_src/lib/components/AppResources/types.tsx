import type { LocalizedString } from 'typesafe-i18n';

import { preferencesText } from '../../localization/preferences';
import { reportsText } from '../../localization/report';
import { resourcesText } from '../../localization/resources';
import type { IR, RA, RR } from '../../utils/types';
import { ensure } from '../../utils/types';
import { icons } from '../Atoms/Icons';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpAppResourceDir, Tables } from '../DataModel/types';
import type { AppResourceMode } from './helpers';

export type AppResourceScope =
  | 'collection'
  | 'discipline'
  | 'global'
  | 'user'
  | 'userType';

export type AppResourceType = {
  readonly tableName: keyof Tables & ('SpAppResource' | 'SpViewSetObj');
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
};

export const appResourceTypes: RR<AppResourceMode, AppResourceType> = {
  appResources: {
    tableName: 'SpAppResource',
    icon: icons.cog,
    label: resourcesText.appResource(),
  },
  viewSets: {
    tableName: 'SpViewSetObj',
    icon: icons.pencilAt,
    label: resourcesText.formDefinitions(),
  },
};

export type ScopedAppResourceDir = SerializedResource<SpAppResourceDir> & {
  readonly scope: AppResourceScope;
};

type AppResourceSubType = {
  readonly mimeType: string | undefined;
  readonly name: string | undefined;
  readonly documentationUrl: string | undefined;
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
  /**
   * Whether when creating a new app resource of this type, should copy the
   * contents from an existing app resource of that type that is in current
   * scope.
   * Default value:
   * If app resource type can only have one specific name, this is true
   * Else false
   */
  readonly useTemplate?: boolean;
  /**
   * Only allow creating this app resource at certain levels
   */
  readonly scope?: RA<AppResourceScope>;
};

/**
 * The order of the subtypes matters. The filtering algorithm loops over these
 * in the order they are defined to find the first subType that matches the
 * current resource. Thus, subtypes should be sorted from the most
 * specific to the least specific.
 */
export const appResourceSubTypes = ensure<IR<AppResourceSubType>>()({
  label: {
    mimeType: 'jrxml/label',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.ticket,
    label: reportsText.label(),
  },
  report: {
    mimeType: 'jrxml/report',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.documentReport,
    label: reportsText.report(),
  },
  userPreferences: {
    mimeType: 'application/json',
    name: 'UserPreferences',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/specify-7-user-preferences-webinar/861',
    icon: icons.cog,
    label: preferencesText.userPreferences(),
    useTemplate: false,
    scope: ['user'],
  },
  defaultUserPreferences: {
    mimeType: 'application/json',
    name: 'DefaultUserPreferences',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Setting-default-user-preferences',
    icon: icons.cog,
    label: preferencesText.defaultUserPreferences(),
  },
  collectionPreferences: {
    mimeType: 'application/json',
    name: 'CollectionPreferences',
    documentationUrl: undefined,
    icon: icons.cog,
    label: preferencesText.collectionPreferences(),
    scope: ['collection'],
  },
  leafletLayers: {
    mimeType: 'application/json',
    name: 'leaflet-layers',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Adding-Custom-Tile-Servers',
    icon: icons.locationMarker,
    label: resourcesText.leafletLayers(),
  },
  rssExportFeed: {
    mimeType: 'text/xml',
    name: 'ExportFeed',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Darwin-Core-Archive-Publishing',
    icon: icons.upload,
    label: resourcesText.rssExportFeed(),
  },
  expressSearchConfig: {
    mimeType: 'text/xml',
    name: 'ExpressSearchConfig',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/simple-search-config/183',
    icon: icons.search,
    label: resourcesText.expressSearchConfig(),
  },
  typeSearches: {
    mimeType: 'text/xml',
    name: 'TypeSearches',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/adding-a-non-native-query-combo-box/859#h-1-type-search-definition-typesearch_defxml-8',
    icon: icons.documentSearch,
    label: resourcesText.typeSearches(),
  },
  webLinks: {
    mimeType: 'text/xml',
    name: 'WebLinks',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/weblinks.xml',
    icon: icons.externalLink,
    label: resourcesText.webLinks(),
  },
  uiFormatters: {
    mimeType: 'text/xml',
    name: 'UIFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/uiformatters.xml',
    icon: icons.hashtag,
    label: resourcesText.uiFormatters(),
  },
  dataObjectFormatters: {
    mimeType: 'text/xml',
    name: 'DataObjFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/dataobj_formatters.xml',
    icon: icons.variable,
    label: resourcesText.dataObjectFormatters(),
  },
  dataEntryTables: {
    mimeType: 'text/xml',
    name: 'DataEntryTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/fish/dataentry_task.xml',
    icon: icons.pencilAt,
    label: resourcesText.dataEntryTables(),
  },
  interactionsTables: {
    mimeType: 'text/xml',
    name: 'InteractionsTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/interactionstask.xml',
    icon: icons.chat,
    label: resourcesText.interactionsTables(),
  },
  otherXmlResource: {
    mimeType: 'text/xml',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.code,
    label: resourcesText.otherXmlResource(),
  },
  otherJsonResource: {
    mimeType: 'application/json',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.adjustments,
    label: resourcesText.otherJsonResource(),
  },
  otherPropertiesResource: {
    mimeType: 'text/x-java-properties',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.viewList,
    label: resourcesText.otherPropertiesResource(),
  },
  otherAppResources: {
    mimeType: undefined,
    name: undefined,
    documentationUrl: undefined,
    icon: icons.document,
    label: resourcesText.otherAppResource(),
  },
} as const);
