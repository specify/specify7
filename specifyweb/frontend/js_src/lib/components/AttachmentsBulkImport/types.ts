import type React from 'react';
import type { LocalizedString } from 'typesafe-i18n';
import type { State } from 'typesafe-reducer';

import type { RA, RR } from '../../utils/types';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import type { Attachment, Tables } from '../DataModel/types';
import type { DatasetBase, DatasetBriefBase } from '../WbPlanView/Wrapped';
import type { PartialAttachmentUploadSpec } from './Import';
import type { staticAttachmentImportPaths } from './importPaths';
import type { keyLocalizationMapAttachment } from './utils';

export type UploadAttachmentSpec = {
  readonly token: string;
  readonly attachmentLocation: string;
};

type Matched = State<
  'matched',
  {
    readonly id: number;
  }
>;

type Skipped = State<
  'cancelled' | 'skipped',
  {
    readonly reason: keyof typeof keyLocalizationMapAttachment;
  }
>;
type Success = State<
  'success',
  {
    readonly successType: 'deleted' | 'uploaded';
  }
>;
export type AttachmentStatus = Matched | Skipped | Success;

export type PartialUploadableFileSpec = Partial<UploadableFileSpec> &
  Pick<UploadableFileSpec, 'uploadFile'>;

type UploadableFileSpec = {
  readonly uploadFile: UnBoundFile;
  readonly matchedId: RA<number>;
  readonly status: AttachmentStatus;
  readonly disambiguated: number;
  readonly attachmentId: number;
  readonly uploadTokenSpec: UploadAttachmentSpec;
  /*
   * This is added because if an attachment was uploaded correctly, but a network error
   * happened while saving the resource, and we retry, we will loose the previous attachment
   */
  readonly attachmentFromPreviousTry?: SpecifyResource<Attachment>;
};

/*
 * Forcing keys to be primitive because objects would be
 * ignored during syncing to backend.
 */
export type BoundFile = Pick<File, 'name' | 'size' | 'type'>;

export type UnBoundFile = {
  readonly file: BoundFile | File;
  readonly parsedName?: string;
};

export type AttachmentWorkProgress = {
  readonly total: number;
  readonly uploaded: number;
  readonly type: 'interrupted' | 'safe' | 'stopped' | 'stopping';
  readonly retryingIn: number;
  readonly stoppedByUser: boolean;
};

export type AttachmentWorkRef = {
  // eslint-disable-next-line functional/prefer-readonly-type
  mappedFiles: RA<PartialUploadableFileSpec>;
  // eslint-disable-next-line functional/prefer-readonly-type
  uploadPromise: Promise<number | undefined>;
  readonly retrySpec: Record<number, number>;
};

export type AttachmentWorkStateProps = {
  readonly onCompletedWork: (
    uploadables: RA<PartialUploadableFileSpec> | undefined
  ) => void;
  readonly workProgress: AttachmentWorkProgress;
  readonly workRef: React.MutableRefObject<AttachmentWorkRef>;
  readonly onStop: () => void;
  readonly triggerNow: () => void;
  readonly dialogText: RR<
    'onAction' | 'onCancelled' | 'onCancelledDescription',
    LocalizedString
  >;
};

export type AttachmentDatasetBrief = DatasetBriefBase & {
  readonly uploaderstatus:
    | 'deleting'
    | 'deletingInterrupted'
    | 'main'
    | 'uploading'
    | 'uploadInterrupted'
    | 'validating';
};

export type AttachmentDataSetPlan = AttachmentDatasetBrief & {
  readonly uploadplan: {
    readonly staticPathKey: string;
  };
};

export type AttachmentDataSet = AttachmentDatasetBrief &
  DatasetBase & {
    readonly rows: RA<PartialUploadableFileSpec>;
    readonly uploadplan: PartialAttachmentUploadSpec;
    readonly uploadresult?: {
      readonly timestamp: string;
    };
  };

export type FetchedDataSet =
  | AttachmentDataSet &
      (
        | { readonly uploaderstatus: 'main' }
        | ({
            readonly uploaderstatus: 'deleting' | 'uploading';
            readonly rows: RA<PartialUploadableFileSpec>;
          } & {
            readonly uploadplan: {
              readonly staticPathKey: keyof typeof staticAttachmentImportPaths;
            };
          })
      );

export type WrappedActionProps<KEY extends keyof Tables> = {
  readonly uploadableFile: PartialUploadableFileSpec;
  readonly baseTableName: KEY;
  readonly dryRun: boolean;
  readonly triggerRetry?: () => void;
};
