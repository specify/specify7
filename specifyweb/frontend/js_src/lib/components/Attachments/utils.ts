import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
