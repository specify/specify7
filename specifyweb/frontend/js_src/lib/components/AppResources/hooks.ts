import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { useAsyncState } from '../../hooks/useAsyncState';
import { f } from '../../utils/functools';
import type { GetOrSet, IR, RA } from '../../utils/types';
import { addMissingFields } from '../DataModel/addMissingFields';
import { fetchCollection } from '../DataModel/collection';
import type { SerializedResource } from '../DataModel/helperTypes';
import type {
  Collection,
  Discipline,
  SpAppResource,
  SpAppResourceData,
  SpecifyUser,
  SpViewSetObj,
} from '../DataModel/types';
import { userPreferences } from '../Preferences/userPreferences';
import { getAppResourceType } from './filtersHelpers';
import { getAppResourceCount, getAppResourceMode } from './helpers';
import { getAppResourceTree, getScope } from './tree';
import type { ScopedAppResourceDir } from './types';
import { appResourceSubTypes } from './types';

export type AppResources = {
  readonly directories: RA<ScopedAppResourceDir>;
  readonly disciplines: RA<SerializedResource<Discipline>>;
  readonly collections: RA<SerializedResource<Collection>>;
  readonly users: RA<SerializedResource<SpecifyUser>>;
  readonly appResources: RA<SerializedResource<SpAppResource>>;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
};

export function useAppResources(
  loadingScreen: boolean = true
): GetOrSet<AppResources | undefined> {
  return useAsyncState(
    React.useCallback(
      async () =>
        f.all({
          directories: fetchCollection('SpAppResourceDir', {
            limit: 0,
            domainFilter: false,
          }).then<AppResources['directories']>(({ records }) =>
            records.map((record) => ({ ...record, scope: getScope(record) }))
          ),
          disciplines: fetchCollection('Discipline', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          collections: fetchCollection('Collection', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          users: fetchCollection('SpecifyUser', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          appResources: fetchCollection('SpAppResource', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          viewSets: fetchCollection('SpViewSetObj', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
        }),
      []
    ),
    loadingScreen
  );
}

export type AppResourcesTree = RA<{
  readonly label: LocalizedString;
  /*
   * A string that would be stable thought the lifecycle of an object.
   * Used to identify a tree node when storing conformation it in localStorage.
   *
   * Directory key is created instead of using id, because app resource
   * needs a way to identify directories that don't yet have IDs (i.e, when
   * creating an app resource for a collection that didn't have any app resources
   * before, the collection won't have any SpAppResourceDir record, thus not
   * directory ID).
   */
  readonly key: string;
  readonly directory: ScopedAppResourceDir | undefined;
  readonly appResources: RA<
    SerializedResource<SpAppResource> & {
      readonly label?: LocalizedString;
    }
  >;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
  readonly subCategories: AppResourcesTree;
}>;

export function useResourcesTree(resources: AppResources): AppResourcesTree {
  const [localize] = userPreferences.use(
    'appResources',
    'appearance',
    'localizeResourceNames'
  );
  return React.useMemo<AppResourcesTree>(() => {
    const tree = getAppResourceTree(resources);
    return localize ? localizeTree(tree) : tree;
  }, [resources, localize]);
}

const localizeTree = (tree: AppResourcesTree): AppResourcesTree =>
  tree.map(({ appResources, subCategories, ...rest }) => ({
    ...rest,
    appResources: appResources.map(localizeResource),
    subCategories: localizeTree(subCategories),
  }));

function localizeResource(
  resource: SerializedResource<SpAppResource> & {
    readonly label?: LocalizedString;
  }
): SerializedResource<SpAppResource> & { readonly label?: LocalizedString } {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  // Check that resource of this type can only have one specific name
  return typeof type.name === 'string'
    ? {
        ...resource,
        // Then replace the name with a localized label unless it's already set
        label: resource.label ?? type.label,
      }
    : resource;
}

export function useAppResourceCount(
  resourcesTree: AppResourcesTree[number]
): number {
  return React.useMemo(
    () => getAppResourceCount(resourcesTree),
    [resourcesTree]
  );
}

/**
 * Fetch resource contents
 */
export function useAppResourceData(
  resource: SerializedResource<SpAppResource | SpViewSetObj>,
  initialData: string | undefined
): {
  readonly resourceData: GetOrSet<
    SerializedResource<SpAppResourceData> | undefined
  >;
  readonly isChanged: boolean;
} {
  const initialValue = React.useRef<string>('');
  const [resourceData, setResourceData] = useAsyncState<
    SerializedResource<SpAppResourceData>
  >(
    React.useCallback(async () => {
      const relationshipName =
        getAppResourceMode(resource) === 'appResources'
          ? 'spAppResource'
          : 'spViewSetObj';
      const newResource = addMissingFields('SpAppResourceData', {
        [relationshipName]: resource.id,
        data: initialData,
      });
      const dataResource =
        typeof resource.id === 'number'
          ? await fetchCollection('SpAppResourceData', {
              limit: 1,
              [relationshipName]: resource.id,
              domainFilter: false,
            }).then(
              ({ records }) =>
                /*
                 * For some reason, app resource can have multiple app resource
                 * datas (but it never does in practice)
                 */
                records[0] ?? newResource
            )
          : newResource;
      const newData = fixLineBreaks(dataResource.data ?? '');
      initialValue.current = newData;
      return { ...dataResource, data: newData };
    }, [resource, initialData]),
    true
  );
  return {
    resourceData: [resourceData, setResourceData],
    isChanged: initialValue.current !== resourceData?.data,
  };
}

const fixLineBreaks = (string: string): string =>
  string.replaceAll(/[\n\r]+/gu, '\n');

export const getAppResourceExtension = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): string =>
  resource._tableName === 'SpViewSetObj'
    ? 'xml'
    : getResourceExtension(resource as SerializedResource<SpAppResource>);

function getResourceExtension(
  resource: SerializedResource<SpAppResource>
): 'jrxml' | 'json' | 'properties' | 'txt' | 'xml' {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  const mimeType = resource.mimeType?.toLowerCase() ?? type?.mimeType ?? '';
  if (mimeType in mimeMapper) return mimeMapper[mimeType];
  else if (mimeType.startsWith('jrxml')) return 'jrxml';
  else if (
    resource.name === 'preferences' &&
    mimeType === appResourceSubTypes.otherPropertiesResource.mimeType
  )
    return 'properties';
  else return 'txt';
}

const mimeMapper: IR<'json' | 'properties' | 'txt' | 'xml'> = {
  'application/json': 'json',
  'text/xml': 'xml',
  'text/plain': 'txt',
};
