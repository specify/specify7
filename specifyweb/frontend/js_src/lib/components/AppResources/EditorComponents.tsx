import { json } from '@codemirror/lang-json';
import { xml } from '@codemirror/lang-xml';
import { indentUnit, StreamLanguage } from '@codemirror/language';
import { properties } from '@codemirror/legacy-modes/mode/properties';
import type { Diagnostic } from '@codemirror/lint';
import { lintGutter, openLintPanel } from '@codemirror/lint';
import { search } from '@codemirror/search';
import type { Extension } from '@codemirror/state';
import { EditorState } from '@codemirror/state';
import { EditorView } from 'codemirror';
import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { useBooleanState } from '../../hooks/useBooleanState';
import { commonText } from '../../localization/common';
import { notificationsText } from '../../localization/notifications';
import { resourcesText } from '../../localization/resources';
import type { RA } from '../../utils/types';
import { filterArray } from '../../utils/types';
import { Button } from '../Atoms/Button';
import { DataEntry } from '../Atoms/DataEntry';
import { LoadingContext } from '../Core/Contexts';
import { getField } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { getFieldBlockerKey, useSaveBlockers } from '../DataModel/saveBlockers';
import type { SpAppResource, SpViewSetObj } from '../DataModel/types';
import { DeleteButton } from '../Forms/DeleteButton';
import { Dialog } from '../Molecules/Dialog';
import { downloadFile, FilePicker, fileToText } from '../Molecules/FilePicker';
import { hasToolPermission } from '../Permissions/helpers';
import { userPreferences } from '../Preferences/userPreferences';
import type { BaseSpec } from '../Syncer';
import type { SimpleXmlNode } from '../Syncer/xmlToJson';
import { jsonLinter, xmlLinter } from './codeMirrorLinters';
import type { getResourceType } from './filtersHelpers';
import { getAppResourceExtension } from './hooks';
import { appResourceSubTypes, appResourceTypes } from './types';

export const appResourceIcon = (
  type: ReturnType<typeof getResourceType>
): JSX.Element => (
  <span
    /*
     * FEATURE: include this title in aria-describedby on the button itself
     *   so that screen readers announce the type after the name
     */
    aria-hidden
    title={
      type === 'viewSet'
        ? resourcesText.formDefinitions()
        : appResourceSubTypes[type].label
    }
  >
    {type === 'viewSet'
      ? appResourceTypes.viewSets.icon
      : appResourceSubTypes[type].icon}
  </span>
);

export function AppResourceEditButton({
  title,
  children,
  appResource,
  onDeleted: handleDeleted,
}: {
  readonly title: LocalizedString;
  readonly children: JSX.Element;
  readonly appResource: SpecifyResource<SpAppResource>;
  readonly onDeleted: (() => void) | undefined;
}): JSX.Element {
  const [isEditingMeta, handleEditingMeta, handleEditedMeta] =
    useBooleanState();
  return (
    <>
      <DataEntry.Edit onClick={handleEditingMeta} />
      {isEditingMeta && (
        <Dialog
          buttons={
            <>
              {!appResource.isNew() &&
              hasToolPermission('resources', 'delete') &&
              typeof handleDeleted === 'function' ? (
                <DeleteButton
                  resource={appResource}
                  onDeleted={handleDeleted}
                />
              ) : undefined}
              <span className="-ml-2 flex-1" />
              <Button.DialogClose>{commonText.close()}</Button.DialogClose>
            </>
          }
          dimensionsKey="AppResourceEdit"
          header={title}
          onClose={handleEditedMeta}
        >
          {children}
        </Dialog>
      )}
    </>
  );
}

export function AppResourceLoad({
  onLoaded: handleLoaded,
}: {
  readonly onLoaded: (data: string, mimeType: string) => void;
}): JSX.Element {
  const [isOpen, handleOpen, handleClose] = useBooleanState();
  const loading = React.useContext(LoadingContext);
  return (
    <>
      <Button.Success className="whitespace-nowrap" onClick={handleOpen}>
        {resourcesText.loadFile()}
      </Button.Success>
      {isOpen && (
        <Dialog
          buttons={commonText.cancel()}
          header={resourcesText.loadFile()}
          onClose={handleClose}
        >
          <FilePicker
            acceptedFormats={undefined}
            onFileSelected={(file): void =>
              loading(
                fileToText(file)
                  .then((data) => handleLoaded(data, file.type))
                  .finally(handleClose)
              )
            }
          />
        </Dialog>
      )}
    </>
  );
}

export function AppResourceDownload({
  resource,
  data,
}: {
  readonly resource: SerializedResource<SpAppResource | SpViewSetObj>;
  readonly data: string;
}): JSX.Element {
  const loading = React.useContext(LoadingContext);
  return (
    <Button.Success
      className="whitespace-nowrap"
      disabled={data.length === 0}
      onClick={(): void =>
        loading(
          downloadFile(
            `${resource.name}.${getAppResourceExtension(resource)}`,
            data
          )
        )
      }
    >
      {notificationsText.download()}
    </Button.Success>
  );
}

export function useIndent(): string {
  const [indentSize] = userPreferences.use(
    'appResources',
    'behavior',
    'indentSize'
  );
  const [indentWithTab] = userPreferences.use(
    'appResources',
    'behavior',
    'indentWithTab'
  );
  return indentWithTab ? '\t' : ' '.repeat(indentSize);
}

/**
 * Use useIndent() instead whenever possible
 */
export function getIndent(): string {
  const indentSize = userPreferences.get(
    'appResources',
    'behavior',
    'indentSize'
  );
  const indentWithTab = userPreferences.get(
    'appResources',
    'behavior',
    'indentWithTab'
  );
  return indentWithTab ? '\t' : ' '.repeat(indentSize);
}

export function useCodeMirrorExtensions(
  resource: SerializedResource<SpAppResource | SpViewSetObj>,
  appResource: SpecifyResource<SpAppResource | SpViewSetObj>,
  xmlSpec: (() => BaseSpec<SimpleXmlNode>) | undefined
): RA<Extension> {
  const [lineWrap] = userPreferences.use(
    'appResources',
    'behavior',
    'lineWrap'
  );
  const [indentSize] = userPreferences.use(
    'appResources',
    'behavior',
    'indentSize'
  );
  const indentCharacter = useIndent();

  const mode = React.useMemo(
    () => getAppResourceExtension(resource),
    [resource]
  );
  const field = React.useMemo(
    () => getField(appResource.specifyTable, 'spAppResourceDatas'),
    [appResource.specifyTable]
  );

  const [extensions, setExtensions] = React.useState<RA<Extension>>([]);
  const [_blockers, setBlockers] = useSaveBlockers(appResource, field);

  React.useEffect(() => {
    let isFirstLint = true;
    function handleLinted(results: RA<Diagnostic>, view: EditorView): void {
      if (isFirstLint && results.length > 0) {
        isFirstLint = false;
        setTimeout(() => {
          const currentFocus = document.activeElement as HTMLElement | null;
          openLintPanel(view);
          currentFocus?.focus();
        }, 0);
      }
      setBlockers(
        filterArray(
          results.map(({ message, severity }) =>
            severity === 'error' ? message : undefined
          )
        ),
        getFieldBlockerKey(field, 'appResourceError')
      );
    }

    const language =
      mode === 'json'
        ? [json(), jsonLinter(handleLinted)]
        : mode === 'properties'
          ? [StreamLanguage.define(properties)]
          : mode === 'jrxml' || mode === 'xml'
            ? [xml(), xmlLinter(xmlSpec?.())(handleLinted)]
            : [];
    setExtensions([
      ...language,
      ...(lineWrap ? [EditorView.lineWrapping] : []),
      indentUnit.of(indentCharacter),
      EditorState.tabSize.of(indentSize),
      lintGutter(),
      search(),
    ]);

    return (): void =>
      setBlockers([], getFieldBlockerKey(field, 'appResourceError'));
  }, [
    appResource,
    mode,
    lineWrap,
    indentCharacter,
    indentSize,
    xmlSpec,
    setBlockers,
  ]);

  return extensions;
}
