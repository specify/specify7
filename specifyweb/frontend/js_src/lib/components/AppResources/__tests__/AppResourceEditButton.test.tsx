import { waitFor } from '@testing-library/react';
import React from 'react';

import { clearIdStore } from '../../../hooks/useId';
import { overrideAjax } from '../../../tests/ajax';
import { requireContext } from '../../../tests/helpers';
import { mount } from '../../../tests/reactUtils';
import { Http } from '../../../utils/ajax/definitions';
import { f } from '../../../utils/functools';
import { localized } from '../../../utils/types';
import { LoadingContext } from '../../Core/Contexts';
import { tables } from '../../DataModel/tables';
import { UnloadProtectsContext } from '../../Router/UnloadProtect';
import { AppResourceEditButton } from '../EditorComponents';
import { testAppResources } from './testAppResources';

requireContext();

beforeEach(() => {
  clearIdStore();
});

jest.setTimeout(1000 * 60 * 60 * 24);

describe('AppResourceEditButton', () => {
  test('simple render', () => {
    const resource = new tables.SpAppResource.Resource({
      name: 'TestName',
      spAppResourceDir: testAppResources.directories[0].resource_uri,
    });

    const { asFragment } = mount(
      <AppResourceEditButton
        appResource={resource}
        title={localized('TestTitle')}
        onDeleted={f.void}
      >
        <></>
      </AppResourceEditButton>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('edit click test', async () => {
    const resource = new tables.SpAppResource.Resource({
      name: 'TestName',
      spAppResourceDir: testAppResources.directories[0].resource_uri,
    });
    const handleDeleted = jest.fn();

    const { getByRole, user } = mount(
      <UnloadProtectsContext.Provider value={[]}>
        <AppResourceEditButton
          appResource={resource}
          title={localized('TestTitle')}
          onDeleted={handleDeleted}
        >
          <></>
        </AppResourceEditButton>
      </UnloadProtectsContext.Provider>
    );

    const button = getByRole('button');
    await user.click(button);

    await waitFor(() => {
      getByRole('dialog');
    });

    const dialog = getByRole('dialog');
    expect(dialog.innerHTML).toMatchInlineSnapshot(`
      "<div class=\\"
                flex items-center gap-2 md:gap-4
                -m-4 cursor-move p-4
                flex-wrap
              \\" id=\\"modal-0-handle\\"><div class=\\"flex items-center gap-2\\"><h2 class=\\"font-semibold text-black dark:text-white text-xl\\" id=\\"modal-0-header\\">TestTitle</h2></div></div><div class=\\"
                  dark:text-neutral-350 -mx-1 flex-1 overflow-y-auto px-1 py-4
                  text-gray-700 flex flex-col gap-2
                \\" id=\\"modal-0-content\\"></div><div class=\\"flex gap-2 justify-end\\"><span class=\\"-ml-2 flex-1\\"></span><button class=\\"button rounded cursor-pointer active:brightness-80 px-4 py-2
          disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
          dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm button hover:brightness-90 dark:hover:brightness-125 bg-[color:var(--secondary-button-color)] text-gray-800 shadow-sm
      dark:text-gray-100\\" type=\\"button\\">Close</button></div>"
    `);

    expect(handleDeleted).not.toHaveBeenCalled();
  });

  overrideAjax(`/delete_blockers/delete_blockers/spappresource/3/`, []);
  overrideAjax(`/api/specify/spappresource/3/`, {
    id: 3,
    name: 'TestName',
    spAppResourceDir: testAppResources.directories[0].resource_uri,
  });
  overrideAjax(`/api/specify/spappresource/3/`, '', {
    method: 'DELETE',
    responseCode: Http.NO_CONTENT,
  });

  test('delete resource test', async () => {
    const resource = new tables.SpAppResource.Resource({
      id: 3,
    });

    const handleDeleted = jest.fn();
    const promiseHandler = jest.fn();

    const { getAllByRole, user, getByRole } = mount(
      <UnloadProtectsContext.Provider value={[]}>
        <LoadingContext.Provider value={promiseHandler}>
          <AppResourceEditButton
            appResource={resource}
            title={localized('TestTitle')}
            onDeleted={handleDeleted}
          >
            <></>
          </AppResourceEditButton>
        </LoadingContext.Provider>
      </UnloadProtectsContext.Provider>
    );

    const button = getByRole('button');
    await user.click(button);

    await waitFor(() => {
      getByRole('dialog');
    });

    await user.click(getAllByRole('button')[1]);

    await user.click(getAllByRole('button')[3]);

    expect(handleDeleted).toHaveBeenCalledTimes(1);
  });
});
