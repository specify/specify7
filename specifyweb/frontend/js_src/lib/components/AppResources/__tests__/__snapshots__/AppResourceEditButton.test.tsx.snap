// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppResourceEditButton simple render 1`] = `
<DocumentFragment>
  <button
    aria-label="Edit"
    class="icon link rounded !text-orange-400 print:hidden "
    title="Edit"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13.586 3.586a2 2 0 112.828 2.828l-.793.793-2.828-2.828.793-.793zM11.379 5.793L3 14.172V17h2.828l8.38-8.379-2.83-2.828z"
      />
    </svg>
  </button>
</DocumentFragment>
`;
